Total Project for Review

Whoistolu opened the pull request:
# Total Project for Review
### On this milestone, we:

- Created and implemented all components in relation to Mission. Below are the details;
- Created a directory for all Redux state slice files.
- Fetched data from the Missions endpoint (https://api.spacexdata.com/v3/missions) when a user navigates to the Missions section.
- Fetched all the rockets data from the RocketsAPI: `https://api.spacexdata.com/v3/rockets`
- Mapped through the data to get the properties values needed id, description, name, and image.
- Once the data were fetched, dispatched an action to store the selected data in Redux store: mission_id, mission_name, description.
- Implemented Missions actions buttons. When a user clicks the "Join Mission" button, an action is dispatched to update the store.
- Implemented the same logic for leaving the mission button.
- Implemented Switch badges for conditional components. Missions that the user has joined already should show a badge "Active Member" instead of the default "NOT A MEMBER" and a button "Leave Mission" instead of the "Join Mission" button (as per design).
- Added styling to my mission table.
- Add my Profile to components to display missions and rockets.

Whoistolu committed:
Set up react app

Whoistolu committed:
Install react-redux, install redux logger, install react router

Whoistolu committed:
Add .github/workflows for linter checks

Whoistolu committed:
Install eslint, babel, stylelint

luqmanmusah committed:
Add components

luqmanmusah committed:
Add routes and Switch

Whoistolu committed:
Fix the Header render error

Whoistolu committed:
Add the space logo

Whoistolu committed:
Create MyProfile

Whoistolu committed:
Merge pull request #26 from Whoistolu/feature/header

Header

Whoistolu committed:
Create action, reducer and store for rockets

Whoistolu committed:
Merge pull request #27 from Whoistolu/rockets-setup

Created directory for Redux state slice files

luqmanmusah committed:
Add active links

luqmanmusah committed:
Add missions actions, reducers and store

luqmanmusah committed:
Add mission.js file

Whoistolu committed:
Fetch dat for rockets from the given API

Whoistolu committed:
Fix github/workflow errors

Whoistolu committed:
Remove console log statement

luqmanmusah committed:
Add missions file to render all missions

Whoistolu committed:
Use useSelector() Redux Hook to select the state slices and render lists of rockets in corresponding routes, Render list of rockets

luqmanmusah committed:
Fix fetch api

Whoistolu committed:
Implement rocket booking actions

luqmanmusah committed:
Add styling to mission page

luqmanmusah committed:
Add styling tpo buttons

Whoistolu committed:
Style of theRocket display page in progress

luqmanmusah committed:
Add styling to conditional statement

Whoistolu committed:
Add extra styles to the rocket page

Whoistolu committed:
Fix styles

Whoistolu committed:
Merge branch 'feature/rockets' into rockets-api

Whoistolu committed:
Merge pull request #28 from Whoistolu/rockets-api

Rockets API

luqmanmusah committed:
Merge branch 'feature/rockets' into mission-setup

luqmanmusah committed:
Merge pull request #29 from Whoistolu/mission-setup

Mission setup

luqmanmusah committed:
Add myProfile file

luqmanmusah committed:
Add changes

Whoistolu committed:
Implement Rocket profile

Whoistolu committed:
Merge branch 'feature/rockets' into profile

luqmanmusah committed:
Add styling to profile

luqmanmusah committed:
Add styling to header

Whoistolu committed:
Myprofile and READNE file

Complete styles for my profile and alsoUpdate the README file with project images

Whoistolu committed:
Update README file

Whoistolu committed:
Update README file

Whoistolu committed:
Update README file

Whoistolu committed:
Fix workflow directory

Whoistolu committed:
Merge pull request #30 from Whoistolu/profile

Profile

Whoistolu committed:
Update README file

Whoistolu committed:
Fix linters errors

Whoistolu committed:
Fix linters errors

Whoistolu committed:
Merge pull request #31 from Whoistolu/feature/rockets

Total Project for Review

Whoistolu committed:
Test

Test mission.js file, my_profile.js, RocketElement.js

Whoistolu committed:
Merge branch 'feature/rockets' into develop

Whoistolu committed:
Updatte README file with image for test

D4nielJ commented on src/components/RocketElement.js:
```diff
@@ -0,0 +1,23 @@
+import SingleRocket from './SingleRockets';
+
+const RocketElement = () => {
+  const dispatch = useDispatch();
+  const rockets = useSelector((state) => state.rockets);
+  useEffect(() => {
+    dispatch(getRockets());
+  }, []);
```
- [x] Please make sure that every time a component renders again, it doesn't fetch rockets or missions if they have already been fetched one time.

```
Actions should be dispatch once
Data should not be added to the store on every re-render (i.e. when changing views / using navigation).
```

![only-one-fetch](https://user-images.githubusercontent.com/79114233/141537596-6b5ba825-7d38-460b-aa9f-6e7ef6df4f92.gif)

Whoistolu replied:
### Hello @D4nielJ ,
Thank you for your feedback.
The issue has just been fixed. Every time a component renders, it doesn't fetch rockets or missions again.

Lameck1 replied:
Done ✔️

D4nielJ commented on src/components/missions.js:
```diff
@@ -0,0 +1,39 @@
+  const state = useSelector((state) => state.missionReducer);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (state.length === 0) {
+      dispatch(loadMisionsThunk());
+    }
+  }, []);
```
- [x] My comment on only fetching data on the first time a component is rendered also applies here.

Whoistolu replied:
Thank you for this.
The issue just got resolved.

Lameck1 replied:
Done ✔️

D4nielJ commented on src/components/missions.js:
```diff
@@ -0,0 +1,39 @@
+      dispatch(loadMisionsThunk());
+    }
+  }, []);
+  console.log(state);
```
- [x] Let's remove any `console.log` statement that we might have in the code. These are good for debugging, but should not be committed to production code.

Whoistolu replied:
Thank you for your feedback.
All console.log statements have been removed.

Lameck1 replied:
Done ✔️

D4nielJ commented on src/components/rocketsApi.js:
```diff
@@ -0,0 +1,12 @@
+    method: 'GET',
+  });
+  const result = await response.json();
+  // console.log(result);
```
- [x] My comment about production code also applies here

Whoistolu replied:
Thank you for your feedback.
All console.log statements have been removed.

Lameck1 replied:
Done ✔️

D4nielJ commented on src/redux/missions/missionReducer.js:
```diff
@@ -0,0 +1,74 @@
+        });
+      dispatch(getMissions(req));
+    } catch (e) {
+      console.log(e);
```
[OPTIONAL] When catching errors you might want to update the state of the application to some `error`, and use this to display some message to the user. Like `There's been an error while loading 'something', you might want to try to reload`

D4nielJ reviewed (changes requested):
## Hello @Whoistolu and @luqmanmusah , 

You're doing great! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights :fire: 
- Redux and React are well structured. :rocket: 
- Your functionality is ready, just minor tweaks on performance are needed.
- Your UI looks good!


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

## Happy coding and stay awesome! 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can always find me on slack!

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Whoistolu committed:
Fix errors as recommended by the reviwer

Whoistolu committed:
Fix Recommendations by Reviewer

Every time a component renders, it doesn't fetch rockets or missions again, All console.log statements have been removed.

Lameck1 reviewed (approved):
Hi, @Whoistolu & @luqmanmusah,

You've done well addressing concerns in the previous review 👏🏼.
Your project is now complete!  Consider merging this branch :shipit: 

Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
